### DISCUSSION TITLE
Mermaid diagrams are rendered randomly

### DISCUSSION THREAD

**Discussion by igeofr** [squidfunk/mkdocs-material#3056] [Q&A] [answered]
Hi,

We have insider status and we try to take full advantage of the mermaid rendering but we get a random rendering. 

``` mermaid
graph LR
  A[Start] --> B{Error?};
  B -->|Yes| C[Hmm...];
  C --> D[Debug];
  D --> B;
  B ---->|No| E[Yay!];
```

If we reload the page several times we get a visual error : 

```
Syntax error in graph
mermaid version 8.8.4
```

And the console shows this error:

```
bundle.e7b111c5.min.js:3 Uncaught Error: Parse error on line 1:
Syntax error in grap
^
Expecting 'open_directive', 'NEWLINE', 'SPACE', 'GRAPH', got 'ALPHA'
    at Xt.parseError (flow.jison:293)
    at Xt.parse (flow.jison:363)
    at xe (flowRenderer.js:291)
    at Object.render (mermaidAPI.js:339)
    at bundle.e7b111c5.min.js:3
    at Object.next (bundle.e7b111c5.min.js:3)
    at t._next (bundle.e7b111c5.min.js:3)
    at t.next (bundle.e7b111c5.min.js:3)
    at bundle.e7b111c5.min.js:3
    at Je (bundle.e7b111c5.min.js:3)
```

The configuration can be viewed here : https://github.com/geotribu/website/blob/features/mermaid/mkdocs.yml#L198-L202

**Comment by squidfunk** [comment]
Thanks for reporting. Can you create a reproducible example? If so, please open a new issue, it’s probably a quick fix then 😊

**Reply by facelessuser** [reply]
Is it several reloads in quick succession? Or is it just several reloads spaced out? Are you using instant mode? Or no instant mode? I could maybe see some kind of race condition if using instant mode with a quick succession of reloads.

I know MathJax has a section in their docs talking about if you are running their process asynchronously: http://docs.mathjax.org/en/latest/web/typeset.html#handling-asynchronous-typesetting. Maybe something similar applies here.

Granted, this is me just dropping some info without knowing exactly how Material is doing what they are doing, but maybe it is helpful.

**Reply by Guts** [reply]
Hi,

As a member of the same team (@geotribu) of @igeofr, I can give some addtional inputs.
This is related to [this PR](https://github.com/geotribu/website/pull/426), where we try to remove the "manual" integration (working quite well in production) of mermaijs in order to take advantage of the out-of-the-box integration of the theme since we are insiders.

> Are you using instant mode? Or no instant mode?

Instant mode is not enabled.

# Two situations can easily be compared

> Can you create a reproducible example?

Yes, using our public repository or comparing the actual situation (in production) with the WIP (on Netlify preview).

To reproduce:

1. Open one of the links below (on preview)
2. If the chart is showing, try to reload the page 2 or 3 times

## Actual setup (production = master)

- [mkdocs.yml configuration](https://github.com/geotribu/website/blob/master/mkdocs.yml#L174-L204):

```yaml
extra_javascript:
  - "theme/assets/javascripts/extra.js"
  - "https://cdn.jsdelivr.net/npm/wa-mediabox@1.0.1/dist/wa-mediabox.min.js"
  - "https://unpkg.com/mermaid@8.6.4/dist/mermaid.min.js"

markdown_extensions:
  - [ ...]
  - pymdownx.superfences:
      custom_fences:
        - name: mermaid
          class: mermaid
          format: !!python/name:pymdownx.superfences.fence_div_format
```

- example working pages:
  1. https://static.geotribu.fr/articles/2021/2021-08-31_pg2datagouv/#pg2datagouv
  2. https://static.geotribu.fr/contribuer/guides/diagrams/#rendu
  3. https://static.geotribu.fr/articles/2020/2020-09-11_html2markdown/#conclusion

![image](https://user-images.githubusercontent.com/1596222/134820525-be54a75d-ef8c-4821-b107-29bd3ff63669.png)

## WIP setup (we use netlify to build previews on our PRs)

- [mkdocs.yml configuration](https://github.com/geotribu/website/blob/features/mermaid/mkdocs.yml#L170-L203):

```yaml
extra_javascript:
  - "theme/assets/javascripts/extra.js"
  - "https://cdn.jsdelivr.net/npm/wa-mediabox@1.0.1/dist/wa-mediabox.min.js"

markdown_extensions:
  - [...]
  - pymdownx.superfences:
      custom_fences:
        - name: mermaid
          class: mermaid
          format: !!python/name:pymdownx.superfences.fence_code_format
```

- same pages, but not working:
  1. https://preview-pullrequest-426--geotribu-preprod.netlify.app/articles/2021/2021-08-31_pg2datagouv/#pg2datagouv
  2. https://preview-pullrequest-426--geotribu-preprod.netlify.app/contribuer/guides/diagrams/
  3. https://preview-pullrequest-426--geotribu-preprod.netlify.app/articles/2020/2020-09-11_html2markdown/#conclusion

![Capture d’écran de 2021-09-26 20-52-09](https://user-images.githubusercontent.com/1596222/134820488-f2a31d09-eed3-488f-9004-ca790c2d4e76.png)

## Differences

- `fence_code_format` is used instead of `fence_div_format`, as [recomended in documentation](https://squidfunk.github.io/mkdocs-material/reference/diagrams/#configuration)
- no more MermaidJS listed as `extra_javascript`

> Tested on Firefox and Chromium on Ubuntu 20.04

**Reply by squidfunk** [reply]
Thanks for the additional input.

> Yes, using our public repository or comparing the actual situation (in production) with the WIP (on Netlify preview).

I'm afraid the repository is too large to start debugging and contains too many customizations. The Netlify preview does not load, i.e., the provided links are not loading – I'm seeing `ERR_CONNECTION_REFUSED`. Could you please try and provide a _small, self-contained reproduction_ I can run locally? Otherwise, it's not possible for me to debug.

This is the first time I'm seeing this behavior. I've verified Material's own [Diagrams page](https://squidfunk.github.io/mkdocs-material/reference/diagrams/#using-sequence-diagrams) which includes Mermaid diagrams and kicks off the loading logic, and I can't reproduce the error you are seeing.

**Reply by squidfunk** [reply]
As a further note: when I'm opening [one of the pages you say are working](https://static.geotribu.fr/articles/2021/2021-08-31_pg2datagouv/#intro), and look into the inspector, I see Mermaid being loaded twice: 8.6.4 (your integration) and 8.8.4 (Insiders). If I refresh the page a few times, sometimes the Mermaid diagram is rendered by Insiders (with matching accent colors), and sometimes it's rendered with the default colors (not adjusting to dark mode). It seem you haven't completely removed your custom integration, which will not work together with Material's automatic integration.

**Reply by Guts** [reply]
> I'm afraid the repository is too large to start debugging and contains too many customizations. 

Ok, I understand.

> The Netlify preview does not load, i.e., the provided links are not loading – I'm seeing ERR_CONNECTION_REFUSED

Strange, it does on my side.

> Could you please try and provide a small, self-contained reproduction I can run locally? Otherwise, it's not possible for me to debug.

I'm not sure having the time to prepare a lighter repo. I'll give it a try.

> This is the first time I'm seeing this behavior. I've verified Material's own Diagrams page which includes Mermaid diagrams and kicks off the loading logic, and I can't reproduce the error you are seeing.

Yes, that's why we're asking here and why we have added the example right from your documentation (only in preview).

> As a further note: when I'm opening one of the pages you say are working, and look into the inspector, I see Mermaid being loaded twice: 8.6.4 (your integration) and 8.8.4 (Insiders). 

Sure! This is the main goal of this: removing manual integration to use only the embedded one. In production, we deployed the insiders version and kept the manual integration because of the described issues. 

This is a part of the problem: it's working (badly but working) when both integrations are used, but not with only the embedded one (in preview).

**Reply by squidfunk** [reply]
Okay, here's what I suspect is happening: 

1. As of your [`mkdocs.yml`](https://github.com/geotribu/website/blob/54e70500a9767140bdbb0702de0efdc05e4c5e64/mkdocs.yml#L204), you're using `fence_div_format`. This is not officially supported with the embedded integration.
2. If you switch to `fence_code_format` and keep the `minify` plugin, all whitespace is stripped from the Mermaid blocks and the syntax becomes invalid, as it is whitespace-sensitive. This is likely why you're seeing the syntax error in the second example.
3. You might try and remove the `minify` plugin for the transitioning phase.

**Reply by Guts** [reply]
> you're using fence_div_format. This is not officially supported with the embedded integration.

Yes, I've switched on the `fence_code_format` on [the switching branch](https://github.com/geotribu/website/blob/features/mermaid/mkdocs.yml#L202)
> You might try and remove the minify plugin for the transitioning phase.

Okay, let's try it!

**Reply by Guts** [reply]
Sorry for the delay, we have been very busy on the tags side lately.

Done:

- switched to `format: !!python/name:pymdownx.superfences.fence_code_format` - see [mkdocs.yml](https://github.com/geotribu/website/blob/master/mkdocs.yml#L208-L212)
- removed the concurrent mermaid.min.js 8.6.4 from th extra_javascript section
- removed the custom script block from the main template ([see this commit](https://github.com/geotribu/website/commit/5af7c7b2eecfb9effc556ea462d3a25d2b98c634))
- try without the minify plugin: no changes. So we enabled it again but disabling JS minifying (see [relevant section into mkdocs.yml](https://github.com/geotribu/website/blob/master/mkdocs.yml#L34-L38))

And... we still have an erratic behavior with sometimes a error message into the FF console:

```console
Uncaught Error: Parse error on line 1:
Syntax error in grap
^
Expecting 'open_directive', 'NEWLINE', 'SPACE', 'GRAPH', got 'ALPHA'
    parseError flow.jison:293
    parse flow.jison:363
    xe flowRenderer.js:291
    render mermaidAPI.js:339
    ui bundle.4b1ed6f7.min.js:3
```

I've made a quick video to show you the behavior:

https://user-images.githubusercontent.com/1596222/137516640-fc6cd49c-2622-470f-ab45-99cbd4ad2465.mp4

**Reply by squidfunk** [reply]
Thanks for getting back. I have no idea where this error is coming from. It certainly doesn't have to do with minifying, so it doesn't really matter how you configure the `minify` plugin. I'm afraid I don't have the time to dig into your code base, but if you can reduce it to a small reproduction, I'm happy to look into it. I haven't seen this behavior until now, and my guess is that it's related to your customizations. You could try to remove your customizations one-by-one to narrow it down.

**Reply by facelessuser** [reply]
I may be able to give it a spin this weekend. I'd like to see how reproducible this is. I'm not currently using Insiders right now, but I can access it and set it up locally to test. I do have Mermaid working with Material without issues: https://facelessuser.github.io/pymdown-extensions/extras/mermaid/.  Material's work I know is at least loosely based on the experiments and work I did, though I honestly haven't looked to see how much (if at all) it deviates.  I may be able to identify what is going on, assuming I can reproduce it.

**Reply by facelessuser** [reply]
Yeah, maybe I'll wait for a minimal example. I too am not excited about sifting through custom conflicts.

**Comment by Guts** [comment, marked as answer]
Hi @igeofr @facelessuser and @squidfunk ,

It seems that it has been fixed by https://github.com/squidfunk/mkdocs-material/issues/3232

**Reply by squidfunk** [reply]
Yep, I was finally able to reproduce it 😊 Glad it's fixed now for you too.
